lib/katello/tasks/clean_candlepin_orphaned_products.rake

```diff
@@ -0,0 +1,61 @@
+require File.expand_path("../engine", File.dirname(__FILE__))
+
+namespace :katello do
+  desc "Runs a katello ping and prints out the statuses of each service"
+  task :check_ping => [:environment, "dynflow:client"] do
+    ::User.current = ::User.anonymous_admin
+    ping_results = Katello::Ping.ping
+    if ping_results[:status] != "ok"
+      pp ping_results
+      fail("Not all the services have been started. Check the status report above and try again.")
+    end
+  end
+
+  desc "Removes custom products (and their associated pools and subscriptions) from Candlepin that are not in Katello"
+  task :clean_candlepin_orphaned_products => ["dynflow:client", "environment"] do
+    User.current = User.anonymous_admin #set a user for orchestration
+    Dir.glob(Katello::Engine.root.to_s + '/app/models/katello/*.rb').each { |file| require file }
+
+    Organization.all.each do |org|
+      print "Cleaning Candlepin orphaned custom products for organization #{org.name}\n"
+
+      deleted_products_count = 0
+      deleted_pools_count = 0
+      deleted_subscriptions_count = 0
+
+      cp_products = Katello::Resources::Candlepin::Product.all(org.label)
+      cp_products = cp_products.select { |prod| Katello::Glue::Candlepin::Product.engineering_product_id?(prod['id']) }
+      cp_product_ids = cp_products.map { |cp_product| cp_product['id'] }
+
+      katello_product_ids = Katello::Product.where(:organization_id => org.id).pluck(:cp_id)
+      orphaned_cp_product_ids = cp_product_ids - katello_product_ids
+      orphaned_cp_product_ids.each do |cp_product_id|
+        
+        pools = Katello::Resources::Candlepin::Pool.get_for_product(cp_product_id, org.label)
+        pools.each do |pool|
+          pool_subscription_id = pool['subscriptionId']
+          if pool_subscription_id.present?
+            print "Deleting Candlepin orphaned subscription #{pool_subscription_id}\n"
+            deleted_subscriptions_count += 1
+            Katello::Resources::Candlepin::Subscription.destroy(pool_subscription_id)
+          end
```

Comment:
Unlike for pools, Candlepin doesn't have an API endpoint that allows you to retrieve subscriptions by product ID. So I noticed that `subscriptionId` is included in the API response [here](https://petstore.swagger.io/?url=https://raw.githubusercontent.com/candlepin/candlepin/master/api/candlepin-api-spec.yaml#/owner/listOwnerPools), and thought we could use that.

However, `subscriptionId` always seems to be nil. I attached my custom product's subscription to a host and to an AK, but it still doesn't find any subscriptions to delete from Candlepin. Maybe we don't need to clean up subscriptions after all..? Or maybe I'm missing something. :thinking:

Reply:
Do we delete subscriptions when deleting custom repository / products? If yes, then I suppose we'd need to support cleaning up orphans. Is there any change that Candlepin detects them as orphans themselves and does its own cleanup?

Reply:
> Do we delete subscriptions when deleting custom repository / products?

you're a genius!!

Just discovered `Actions::Candlepin::Product::DeletePools` and `Actions::Candlepin::Product::DeleteSubscriptions` - turns out I've been trying to reinvent these and they were there all along. Now my rake task can just run those! I can just pass in the product's cp_id.

Reply:
Hey awesome!